src/v/hashing/utils.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "utils/functional.h"
+
+#include <absl/hash/hash.h>
+#include <boost/functional/hash.hpp>
+
+#include <functional>
+
+namespace hash {
+namespace impl {
+template<typename T>
+concept is_absl_hashable = requires(T const& t) {
+    { absl::Hash<T>{}(t) } -> std::convertible_to<std::size_t>;
+};
+
+template<typename T>
+concept is_std_hashable = requires(T const& t) {
+    { std::hash<T>{}(t) } -> std::convertible_to<std::size_t>;
+};
+
+template<typename T>
+concept is_boost_hashable = requires(T const& t) {
+    { boost::hash<T>{}(t) } -> std::convertible_to<std::size_t>;
+};
+
+template<typename T>
+size_t combine(size_t& seed, T const& t) {
+    if constexpr (is_std_hashable<T>) {
+        boost::hash_combine(seed, std::hash<T>{}(t));
+    } else if constexpr (is_absl_hashable<T>) {
+        boost::hash_combine(seed, absl::Hash<T>{}(t));
+    } else if constexpr (is_boost_hashable<T>) {
+        boost::hash_combine(seed, boost::hash<T>{}(t));
+    } else {
+        static_assert(always_false_v<T>, "No hasher found for T");
+    }
+    return seed;
+}
+
+template<typename T>
+size_t combine(size_t& seed, std::optional<T> const& t) {
+    if (!t.has_value()) {
+        // Arbitray value for empty optional, taken from absl::hash
+        boost::hash_combine(seed, 0x12345678);
+        return seed;
+    }
+    return combine<T>(seed, *t);
+}
```

Comment:
This isn't needed (at least for clang/libc++)

Reply:
IMO keeping it is fine as logic stays the same. Otherwise will need to add conditional logic to only execute when T is a specialization of optional and the type is hashable by boost, is_std_optional is within `v::reflection`, don't want to have hashing depend on that either

Reply:
Isn't this sufficient? https://en.cppreference.com/w/cpp/utility/optional/hash

Reply:
Removed method entirely, thanks Ben